File: RETOS/Reto25/Reto25.java
package RETOS.Reto25;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.Scanner;

/*
 * Crea un programa que simule un evento deportivo de los Juegos Olímpicos,
 * donde el usuario puede crear un evento, agregar participantes, simular la
 * competencia y generar un informe con el resultado, incluyendo el podio y 
 * la clasificación por países.
 */
class Participante {
    String nombre;
    String pais;

    Participante(String nombre, String pais) {
        this.nombre = nombre;
        this.pais = pais;
    }
}

class Evento {
    String nombre;
    ArrayList<Participante> participantes;

    Evento(String nombre) {
        this.nombre = nombre;
        this.participantes = new ArrayList<>();
    }

    void agregarParticipante(String nombre, String pais) {
        participantes.add(new Participante(nombre, pais));
    }

    ArrayList<Participante> simularEvento() {
        if (participantes.size() < 3) {
            System.out.println("Debe haber al menos 3 participantes para simular el evento.");
            return null;
        }
        Collections.shuffle(participantes);
        return new ArrayList<>(participantes.subList(0, 3));
    }

    void generarInforme(ArrayList<Participante> resultados) {
        System.out.println("");
        System.out.println("Simulación del evento '" + nombre + "':");
        System.out.println("🥇 Oro: " + resultados.get(0).nombre + " (" + resultados.get(0).pais + ")");
        System.out.println("🥈 Plata: " + resultados.get(1).nombre + " (" + resultados.get(1).pais + ")");
        System.out.println("🥉 Bronce: " + resultados.get(2).nombre + " (" + resultados.get(2).pais + ")");
    }
}

public class Reto25 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        ArrayList<Evento> eventos = new ArrayList<>();
        Map<String, int[]> rankingPaises = new HashMap<>();

        while (true) {
            System.out.println("\nOpciones:");
            System.out.println("1. Nuevo evento");
            System.out.println("2. Salir");
            String opcion = sc.nextLine();

            if (opcion.equals("1")) {
                System.out.print("Ingresa el nombre del evento: ");
                String nombreEvento = sc.nextLine();
                Evento evento = new Evento(nombreEvento);

                while (true) {
                    System.out.print("Ingresa el nombre del participante (o 0 para terminar): ");
                    String nombreParticipante = sc.nextLine();
                    if (nombreParticipante.equals("0")) {
                        if (!evento.participantes.isEmpty()) {
                            System.out.println("");
                            System.out.println("Participantes registrados:");
                            for (Participante p : evento.participantes) {
                                System.out.println("- " + p.nombre + " (" + p.pais + ")");
                            }
                        }
                        break;
                    }
                    System.out.print("Ingresa el país del participante: ");
                    String pais = sc.nextLine();
                    evento.agregarParticipante(nombreParticipante, pais);
                }

                if (evento.participantes.size() >= 3) {
                    ArrayList<Participante> resultados = evento.simularEvento();
                    if (resultados != null) {
                        evento.generarInforme(resultados);

                        for (int i = 0; i < 3; i++) {
                            String pais = resultados.get(i).pais;
                            rankingPaises.putIfAbsent(pais, new int[3]);
                            if (i == 0) {
                                rankingPaises.get(pais)[0]++;
                            } else if (i == 1) {
                                rankingPaises.get(pais)[1]++;
                            } else {
                                rankingPaises.get(pais)[2]++;
                            }
                        }
                        System.out.println("");
                        System.out.println("\nRanking de países por medallas:");
                        for (Map.Entry<String, int[]> entry : rankingPaises.entrySet()) {
                            String pais = entry.getKey();
                            int[] medallas = entry.getValue();
                            System.out.println(pais + ": " + medallas[0] + " Oro, " + medallas[1] + " Plata, " + medallas[2] + " Bronce");
                        }
                    }
                } else {
                    System.out.println("No hay suficientes participantes para simular el evento (se requieren al menos 3).");
                }
            } else if (opcion.equals("2")) {
                System.out.println("¡Gracias por participar en los Juegos Olímpicos de París 2024!");
                break;
            } else {
                System.out.println("Opción no válida.");
            }
        }
        sc.close();
    }
}

File: RETOS/Reto25/Reto25.js
/*
 * Crea un programa que simule un evento deportivo de los Juegos Olímpicos,
 * donde el usuario puede crear un evento, agregar participantes, simular la
 * competencia y generar un informe con el resultado, incluyendo el podio y 
 * la clasificación por países.
 */

const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

class Evento {
  constructor(nombre) {
    this.nombre = nombre;
    this.participantes = [];
  }

  agregarParticipante(nombre, pais) {
    this.participantes.push({ nombre, pais });
  }

  simularEvento() {
    if (this.participantes.length < 3) {
      console.log("Debe haber al menos 3 participantes para simular el evento.");
      return null;
    }
    this.participantes.sort(() => Math.random() - 0.5);
    return this.participantes.slice(0, 3);
  }

  generarInforme(resultados) {
    console.log(`Simulación del evento '${this.nombre}':`);
    console.log(`🥇 Oro: ${resultados[0].nombre} (${resultados[0].pais})`);
    console.log(`🥈 Plata: ${resultados[1].nombre} (${resultados[1].pais})`);
    console.log(`🥉 Bronce: ${resultados[2].nombre} (${resultados[2].pais})`);
  }
}

const rankingPaises = {};

function mostrarMenu() {
  console.log("\nOpciones:");
  console.log("1. Nuevo evento");
  console.log("2. Salir");
  rl.question("Selecciona una opción: ", (opcion) => {
    if (opcion === "1") {
      rl.question("Ingresa el nombre del evento: ", (nombreEvento) => {
        const evento = new Evento(nombreEvento);

        function agregarParticipante() {
          rl.question("Ingresa el nombre del participante (o 0 para terminar): ", (nombreParticipante) => {
            if (nombreParticipante === "0") {
              if (evento.participantes.length > 0) {
                console.log("\nParticipantes registrados:");
                evento.participantes.forEach((p) => {
                  console.log(`- ${p.nombre} (${p.pais})`);
                });
              }
              return simularEvento();
            }
            rl.question("Ingresa el país del participante: ", (pais) => {
              evento.agregarParticipante(nombreParticipante, pais);
              agregarParticipante();
            });
          });
        }

        function simularEvento() {
          if (evento.participantes.length < 3) {
            console.log("Debe haber al menos 3 participantes para simular el evento.");
            return mostrarMenu();
          }
          const resultados = evento.simularEvento();
          if (resultados) {
            evento.generarInforme(resultados);
            resultados.forEach((participante, index) => {
              const pais = participante.pais;
              if (!rankingPaises[pais]) rankingPaises[pais] = [0, 0, 0];
              rankingPaises[pais][index]++;
            });
            console.log("\nRanking de países por medallas:");
            Object.entries(rankingPaises).forEach(([pais, medallas]) => {
              console.log(`${pais}: ${medallas[0]} Oro, ${medallas[1]} Plata, ${medallas[2]} Bronce`);
            });
          }
          mostrarMenu();
        }

        agregarParticipante();
      });
    } else if (opcion === "2") {
      console.log("¡Gracias por participar en los Juegos Olímpicos de París 2024!");
      rl.close();
    } else {
      mostrarMenu();
    }
  });
}

mostrarMenu();
